spring-cloud-dataflow-module-deployers/spring-cloud-dataflow-module-deployer-kubernetes/src/main/java/org/springframework/cloud/dataflow/module/deployer/kubernetes/KubernetesModuleDeployer.java

```diff
@@ -0,0 +1,306 @@
+ * 
+ * @author Florian Rosenberg
+ */
+public class KubernetesModuleDeployer implements ModuleDeployer {
+	
+	protected static final String SCSM_GROUP_KEY = "scsm-group";
+	protected static final String SCSM_LABEL_KEY = "scsm-label";
+	private static final String SCSM_EXTENSION = "scsm-extension";
+	private static final String SCSM_VERSION = "scsm-version";
+	private static final String SCSM_GROUP_ID = "scsm-groupId";
+	private static final String SCSM_ARTIFACT_ID = "scsm-artifactId";
+	private static final String SPRING_MARKER_VALUE = "scsm-module";
+	private static final String SPRING_MARKER_KEY = "role";
+	private static final String PORT_KEY = "port";
+	private static final String SERVER_PORT_KEY = "server.port";
+	private static final String HEALTH_ENDPOINT = "/health";
+	private static final String CONTAINER_NAME = "spring-module-launcher";
+	
+	private static final String SPRING_REDIS_HOST = "SPRING_REDIS_HOST";
+	private static final String SPRING_REDIS_SENTINEL_NODES = "SPRING_REDIS_SENTINEL_NODES";
+	private static final String SPRING_REDIS_SENTINEL_MASTER = "SPRING_REDIS_SENTINEL_MASTER";
+		
+	/** 
+	 * Use the default spring image. Override with --kubernetes.moduleLauncherImage 
+	 * as a Spring Admin parameter to use a different one.
+	 */
+	private static final String DEFAULT_IMAGE_NAME = "springcloud/stream-module-launcher";
+
+	private final Logger logger = LoggerFactory.getLogger(this.getClass());
+	
+	@Autowired
+	private KubernetesClient kubeClient;
+	
+	private KubernetesModuleDeployerProperties properties;
+
+	public KubernetesModuleDeployer(
+			KubernetesModuleDeployerProperties properties) {
+		this.properties = properties;
+	}
+
+	@Override
+	public ModuleDeploymentId deploy(ModuleDeploymentRequest request) {			
+		ModuleDeploymentId id = ModuleDeploymentId.fromModuleDefinition(request.getDefinition()); 
+		
+		logger.debug("Deploying module: {}", createKubernetesName(id));
+		
+		Integer externalPort = 8080;
+		// we also create a service in case have an '--port' or '--server.port' argument (source or sink)
+		Map<String, String> parameters = request.getDefinition().getParameters();
+		if (parameters.containsKey(PORT_KEY)) {
+			externalPort = Integer.valueOf(parameters.get(PORT_KEY));
+			createService(id, request, externalPort);
+		} else if (parameters.containsKey(SERVER_PORT_KEY)) {
+			externalPort = Integer.valueOf(parameters.get(SERVER_PORT_KEY));		
+			createService(id, request, externalPort);
+		}		
+		createReplicationController(id, request, externalPort);				
+		return id;	
+	}
+
+	@Override
+	public void undeploy(ModuleDeploymentId id) {
+		String name = createKubernetesName(id);
+		logger.debug("Undeploying module: {}", name);
+
+		try {
+			kubeClient.replicationControllers().withName(name).delete();		
+			kubeClient.services().withName(name).delete();
+		} catch (KubernetesClientException e) {
+			logger.error(e.getMessage(), e);
+			// FIXME what do we throw??
```

Comment:
Any RuntimeException would be fine, for now. Need to throw something though, obviously :)